Implement Sentry instrumentation for performance and error tracking

cyrilS-dev opened the pull request:
This Pull Request introduces Sentry integration for performance tracking and error monitoring.

**Key changes:**

1. FrameProcessor:
 
- Add conditional Sentry import and initialization check
- Implement custom instrumentation with Sentry spans in FrameProcessorMetrics to measure TTFB (Time To First Byte) and processing time when Sentry is available
- Maintain existing metrics functionality with MetricsFrame regardless of Sentry availability

2. DeepgramSTTService:

     1. Enable general metrics generation:

        - Implement `can_generate_metrics` method, returning True when VAD is enabled
        - This allows metrics to be collected and used by both Sentry and the metrics system in frame_processor.py

     2. Integrate Sentry-compatible performance tracking:

        - Add start_ttfb_metrics and start_processing_metrics calls in the VAD speech detection handler
        - Implement stop_ttfb_metrics call when receiving transcripts
        - Add stop_processing_metrics for final transcripts

     3. Enhance VAD support for metrics:

        - Add `vad_enabled` property to check VAD event availability
        - Implement VAD-based speech detection handler for precise metric timing


These changes allow for comprehensive performance monitoring and analysis using Sentry, providing valuable insights into STT, LLM and TTS services. The integration is designed to be non-intrusive, maintaining existing functionality while adding custom performance instrumentation.

**Testing:**

To validate the Sentry integration, the following steps were taken:

1. Sentry initialization was added to `examples/simple-chatbot/demo.py`:

```python
sentry_sdk.init(
dsn="SENTRY_DSN",
integrations=[AsyncioIntegration()],
disabled_integrations=[HttpxIntegration()],
traces_sample_rate=1.0,
enable_tracing=True,
)
```
2. A Sentry transaction was created to encompass the main execution of the chatbot:

```python
with sentry_sdk.start_transaction(op="pipecat", name="Pipecat"):
      await runner.run(task)
```
3. The `vad_events` parameter in the DeepgramSTTService configuration was set to `True` 

4. The chatbot was run with this configuration, and the results were observed in the Sentry dashboard.

5. A screenshot of the Sentry performance monitoring results, demonstrating successful tracking of STT, LLM, and TTS services performance metrics :

<img width="526" alt="sentry" src="https://github.com/user-attachments/assets/e553a69d-dbbf-4700-b8e4-0052584fb3bc">

These tests confirm that Sentry is correctly integrated and capable of monitoring both performance and errors across the pipeline components.

**Notes:**

 It's crucial to initialize Sentry as early as possible in the application, and before `from pipecat.processors.frame_processor` : 

- To capture any potential errors that may occur during the imports
- To ensure that `sentry_sdk.is_initialized()` returns the correct state when called in `frame_processor.py`.

**Future Improvements:**

1. Usage Metrics in Sentry Spans:
   I  plan to enhance the Sentry integration by adding "usage_metrics" to Sentry spans as tags. 
   This improvement will provide more detailed and contextual information about resource usage within each service, allowing for better analysis and optimization of our pipeline components.

2. Review TTS Processing Time Measurement:
   I need to review and potentially refine the way we measure processing time for the Text-to-Speech (TTS) service.

aconchillo commented:
This is fantastic. Thank you! I was thinking that in the future we might want to use something different than Sentry.

What if we create a `FrameProcessorMetrics` subclass (inside a new `processors.metrics` module), for example `SentryMetrics`. And then we pas that to the `FrameProcessor` constructor:

```
class FrameProcessor:

    def __init__(
            self,
            *,
            name: str | None = None,
            metrics: FrameProcessorMetrics = FrameProcessorMetrics(),
            loop: asyncio.AbstractEventLoop | None = None,
            **kwargs):
```

cyrilS-dev commented:
Thank you for your feedback @aconchillo ! I really like your suggestion. It's definitely a more modular and scalable approach. I'll start working on this asap!

cyrilS-dev commented:
@aconchillo I've updated the PR to implement a flexible metrics system as you proposed. Here are the key changes:

1. Created a new processors/metrics/ directory with `base.py` containing the `FrameProcessorMetrics` base class and `sentry.py` with the `SentryMetrics` subclass.

2. Modified the `FrameProcessor.__init__` method to accept a `metrics` parameter that can be either `FrameProcessorMetrics` or one of its subclasses

Now, you can choose the metrics implementation for each processor in the pipeline. For example : 

```python
from pipecat.processors.metrics.sentry import SentryMetrics

tts = CartesiaTTSService(
api_key=os.getenv("CARTESIA_API_KEY"),
model_id="sonic-multilingual",
voice_id="65b25c5d-ff07-4687-a04c-da2f43ef6fa9",
language="fr",
metrics=SentryMetrics
)
```

If the `metrics` parameter is not specified, the default `FrameProcessorMetrics` will be used.

aconchillo commented on src/pipecat/processors/frame_processor.py:
```diff
@@ -28,68 +27,13 @@ class FrameDirection(Enum):
             self,
             *,
             name: str | None = None,
+            metrics: FrameProcessorMetrics = FrameProcessorMetrics,
```
I think this should be `FrameProcessorMetrics()` instead of passing a class. It's possible that other metrics implementation need additional arguments (api keys, user name, etc). If we hide the construction we would not be able to do that.

What we can do (since we need the name) is add a `FrameProcessorMetrics.set_processor_name()`:

```
self._metrics = metrics
metrics.set_processor_name(self.name)
```

cyrilS-dev replied:
@aconchillo with `metrics: FrameProcessorMetrics = FrameProcessorMetrics()`, the` FrameProcessorMetrics()` instance is created once when the `FrameProcessor` class is defined. As a result, every time I create a new FrameProcessor, it uses the same instance of FrameProcessorMetrics, and it's a problem. My idea was to use `metrics: FrameProcessorMetrics = None` and 
```python
self._metrics = metrics or FrameProcessorMetrics() 
self._metrics.set_processor_name(self.name)
``` 
 to create a new instance of `FrameProcessorMetrics` inside the constructor if metrics is not provided. Let me know if i'm missing something, thanks.

aconchillo replied:
Oh yes! You are totally right :sweat_smile:  . Well, the important point was for people to be able to create instances from outside which we now can do! :smile:

aconchillo commented:
Looks great! There was a bigger metrics PR that got merged yesterday so unfortunately you will need to rebase your changes. Sorry :-(.

Btw, I think instead of `base.py` we can call it `frame_processor_metrics.py` we do that with other base classes as well.

cyrilS-dev commented:
> Looks great! There was a bigger metrics PR that got merged yesterday so unfortunately you will need to rebase your changes. Sorry :-(.
> 
> Btw, I think instead of `base.py` we can call it `frame_processor_metrics.py` we do that with other base classes as well.

done :white_check_mark:

aconchillo commented on src/pipecat/processors/metrics/sentry.py:
```diff
@@ -0,0 +1,52 @@
+    import sentry_sdk
+    sentry_available = sentry_sdk.is_initialized()
+    if not sentry_available:
+        logger.debug("Sentry SDK not initialized. Sentry features will be disabled.")
```
it seems this should be a warning.

cyrilS-dev replied:
done ✅

aconchillo commented on src/pipecat/processors/metrics/sentry.py:
```diff
@@ -0,0 +1,52 @@
+        logger.debug("Sentry SDK not initialized. Sentry features will be disabled.")
+except ImportError:
+    sentry_available = False
+    logger.debug("Sentry SDK not installed. Sentry features will be disabled.")
```
same here.

cyrilS-dev replied:
done ✅

aconchillo commented:
LGTM! I added a couple of minor comments. Before I merge, do you mind fixing linting issues? Just run autopep8. We will be moving to Ruff shortly, but for now that's what we are checking.

cyrilS-dev commented:
> LGTM! I added a couple of minor comments. Before I merge, do you mind fixing linting issues? Just run autopep8. We will be moving to Ruff shortly, but for now that's what we are checking.

Fixed! Thanks for the help @aconchillo

aconchillo commented:
Merging. Thank you!

cyrilS-dev commented:
thank you @aconchillo !
Just to understand, why the "Formatting lints" job failed?

aconchillo commented:
> thank you @aconchillo ! Just to understand, why the "Formatting lints" job failed?

I just saw... not sure. I'll take a look and fix. Btw, I'll enable `vad_events` by default otherwise we won't have metrics for Deepgram.

cyrilS-dev commented:
> > thank you @aconchillo ! Just to understand, why the "Formatting lints" job failed?
> 
> I just saw... not sure. I'll take a look and fix. Btw, I'll enable `vad_events` by default otherwise we won't have metrics for Deepgram.

I didn't enable `vad_events` by default because it is currently only used for metrics. I'm working on a version where Deepgram `vad_events` emits `UserStartedSpeakingFrame` and `UserStoppedSpeakingFrame` when Silero VAD is disabled. We can discuss this in the dedicated PR that I will create :)

aconchillo commented:
> thank you @aconchillo ! Just to understand, why the "Formatting lints" job failed?

It's because of this. https://github.com/pipecat-ai/pipecat/commit/c7ff79a652f432bb466b8ae7620906ba16b5ea2b

But I just merged the Ruff PR. `autopep8` was doing very weird stuff.

fatwang2 commented:
hey, I have read the doc about how to use sentry in pipecat, but I am still confused, can you share more examples about that?

fatwang2 commented:
it worked on tts, but didn't work on llm
